### DISCUSSION TITLE
California | Upgrade v15.02 to v17 Popup menu bugs when using in p-table

### DISCUSSION THREAD

**Discussion by nazrulcsebd** [primefaces/community#1248] [PrimeNG]
Hi! When upgrade v15.0.2 to v17. Everything looks good, but the Popup menu does not work in p-table. The popup menu renders frequently and the click event does not work in one click, it needs double click. If I initiate the menu items on ngOnInit as static items then it works fine. But, if I use a function to generate the menu items based on table row data then the problem occurs.
 
I'm using the California template (Also updated it to v16.9.0). What can I do now, have any suggestions or some upgrade guide or considerations for my context upgrade?

I give the sample code for dynamic binding of menu items.

`
<p-table id="tblAppPackage" [value]="packageGridObj" styleClass="p-datatable-gridlines">
    <ng-template pTemplate="header">
        <tr>
            <th style="width: 5%">Sr. No.</th>           
            <th style="width: 8%">Status</th>
            <th style="width: 6%">Action</th>
        </tr>
    </ng-template>
    <ng-template pTemplate="body" let-package let-ri="rowIndex">
        <tr>
            <td class="text-center" style="width: 5%;">{{ startRowIndex + ri + 1 }}</td>            
            <td style="width: 8%" class="text-center">
                <span *ngIf="checkPackageStatus(package?.status)" style="color: red; font-weight: bold;">{{
                    checkPackageStatusValue(package) }}</span>
                <span *ngIf="!checkPackageStatus(package?.status)">{{ package?.status }}</span>
            </td>
            <td class="text-center" style="width: 6%">
                <p-menu #menu [popup]="true" [model]="getPackageOptionMenuItem(package)" appendTo="body"></p-menu>
                <button type="button" pButton icon="pi pi-chevron-down" label="Options"
                    (click)="menu.toggle($event)" style="width:auto"></button>
            </td>
        </tr>
    </ng-template>
</p-table>
`

The TS file code for dynamic menu items is based on row data.

`
getPackageOptionMenuItem(packageItem: any) {
    let items: MenuItem[] = [];

    let currPackageSideLink = this.packageSideLinkObj?.filter(pksl => pksl.packageId == packageItem.packageId)[0];

    if (currPackageSideLink) {

      if (currPackageSideLink.processFileInWkf) {
        let processFileInWorkflow: MenuItem = {
          label: 'Process file in Workflow', id: 'linkProcessFileWorkflow', icon: 'pi pi-folder-open', command: (event) => {
            this.fileProcessingForAppWorkflow();
          }
        };
        items.push(processFileInWorkflow);
      }

      if (currPackageSideLink.viewPkg) {
        let viewPackage: MenuItem = {
          label: 'View', id: 'linkViewPackageDetails', icon: 'pi pi-eye', command: (event) => {
            this.goToPackageDetails(packageItem);
          }
        };
        items.push(viewPackage);
      }

      if (currPackageSideLink.publish) {
        let appPackagePublish: MenuItem = {
          label: 'Publish', id: 'linkAppPackagePublish', icon: 'pi pi-check-circle', command: (event) => {
            this.goToAppPackagePublish(packageItem);
          }
        };
        items.push(appPackagePublish);
      }

      
    }

    return items;

  }
`
I inspect the browser when dynamically generating the menu options. Here is the snapshot,

![menu load frequently on p table angular 17](https://github.com/primefaces/community/assets/3226667/7d27bc25-aa29-4310-8e2b-d16ae7b60b55)

But when given static menu items it works fine.

the code is given below 

`
<p-table id="tblAppPackage" [value]="packageGridObj" styleClass="p-datatable-gridlines">
    <ng-template pTemplate="header">
        <tr>
            <th style="width: 5%">Sr. No.</th>           
            <th style="width: 8%">Status</th>
            <th style="width: 6%">Action</th>
        </tr>
    </ng-template>
    <ng-template pTemplate="body" let-package let-ri="rowIndex">
        <tr>
            <td class="text-center" style="width: 5%;">{{ startRowIndex + ri + 1 }}</td>            
            <td style="width: 8%" class="text-center">
                <span *ngIf="checkPackageStatus(package?.status)" style="color: red; font-weight: bold;">{{
                    checkPackageStatusValue(package) }}</span>
                <span *ngIf="!checkPackageStatus(package?.status)">{{ package?.status }}</span>
            </td>
            <td class="text-center" style="width: 6%">
                <p-menu #menu [popup]="true" [model]="itemsOptions" appendTo="body"></p-menu>
                <button type="button" pButton icon="pi pi-chevron-down" label="Options"
                    (click)="menu.toggle($event)" style="width:auto"></button>
            </td>
        </tr>
    </ng-template>
</p-table>
`

The TS file code is given below

`
itemsOptions: MenuItem[] = [];

ngOnInit(): void {

    this.itemsOptions = [
      {
          label: 'Save', icon: 'pi pi-fw pi-check'
      },
      {
          label: 'Update', icon: 'pi pi-fw pi-refresh'
      },
      {
          label: 'Delete', icon: 'pi pi-fw pi-trash'
      },
      {
          separator: true
      },
      {
          label: 'Home', icon: 'pi pi-fw pi-home'
      },
  ];

  }

`

I give the full code to the git repository (https://github.com/nazrulcsebd/california-theme-problem-grid). Please solve the problem as soon as possible.  Please run the code and go to the page URL http://localhost:4200/#/uikit/table

The code link [https://github.com/nazrulcsebd/california-theme-problem-grid](url)

**Comment by cetincakiroglu** [comment]
Hi,

Thanks for the explanation and example. We'll investigate the issue.

@mehmetcetin01140

**Reply by nazrulcsebd** [reply]
When inspecting to browser then the Popup menu does not render correctly. When double click any menu link then it working. We need to update the version as early as possible. please give us a solution

![image](https://github.com/primefaces/community/assets/3226667/62baad91-027b-42a4-b6e7-979b0055c639)

**Reply by cetincakiroglu** [reply]
Hi,

For the render, as you see there is nothing rendered in p-menu tags, all items and overlay are appended before the closing body tag of html. This is not a bug. For the frequent render issue, when we check the dom it only renders for once, however we found that change detection is triggered so many times, we'll try to address this issue within v17.  When it comes to double click issue in the menu, it's related to PrimeNG core, first click invokes the focus method in menu, second click invokes the item command and this is a bug. However, we've downloaded your repo and managed to provide a workaround for the issue you're having until we fix it for good. Could you please follow the steps below:

1-) Modify the HTML

```
 <span *ngIf="col.field === 'peActions' ">
   <button type="button" pButton class="p-button p-button-sm me-2 action-menu" (click)="openMenu(menu, $event, rowData)">
	<span class="ui-button-icon-left pi pi-ellipsis-v"></span>
        <span class="ui-button-icon-left pi pi-angle-down"></span>
   </button>
   <p-menu kay="m{{ri + 1}}" appendTo="body" #menu [model]="menuItems"
       [styleClass]="boxWidth" [popup]="true"></p-menu>
</span>
```

2-) in tabledemo.ts add the following:

```
    dataRow: any;

    menuItems: any[];

    getPeOptionMenuItem() {
        let items: MenuItem[] = [];

        let processComplaint: MenuItem = {
            label: "Process",
            id: "processComplaint",
            disabled: !this.dataRow.process,
            icon: "pi pi-caret-right",
            command: (event) => {
                this.processComplaint(this.dataRow);
            },
        };
        items.push(processComplaint);

        let complaintView: MenuItem = {
            label: "View",
            id: "linkComplaintView",
            disabled: !this.dataRow.view,
            icon: "pi pi-eye",
            command: () => {
                this.goToComplaintView(this.dataRow);
            },
        };
        items.push(complaintView);
        let notificationFromCPTU: MenuItem = {
            label: "Notification from CPTU",
            id: "linkEscalateComplaintNextLevel",
            disabled: !this.dataRow.notificationFromCptu,
            icon: "pi pi-caret-right",
            command: (event) => {
                this.goToNotificationFromCPTUPage(this.dataRow);
            },
        };
        items.push(notificationFromCPTU);

        // console.log("items->", items);
        return items;
    }

    openMenu(menu, event, rowData) {
        this.dataRow = rowData;
        this.menuItems = this.getPeOptionMenuItem();
        menu.show(event);
    }

```

It should work as it is supposed to, could you please try it and give us a feedback?

**Reply by nazrulcsebd** [reply]
Thanks for your solution. The way you provided solved the issue. But we need to change various components.  Please inform us when the **PrimeNG CORE** is update
